discovery_engine_core/core/src/mab.rs

```diff
@@ -46,113 +58,115 @@ impl BetaSample for BetaSampler {
     fn alpha(&self) -> f32;
+
     /// Returns the beta parameter of the beta distribution.
     fn beta(&self) -> f32;
-    /// Returns `true` if the bucket contains no elements.
+
+    /// Checks if the bucket is empty.
     fn is_empty(&self) -> bool;
+
     /// Removes the next best element from this bucket and returns it, or `None` if it is empty.
     fn pop(&mut self) -> Option<T>;
 }
 
-fn pull_arms<B, T>(beta_sampler: &impl BetaSample, buckets: &mut [&mut B]) -> Result<T, Error>
-where
-    B: Bucket<T>,
-{
-    let sample_from_bucket = |bucket: &B| beta_sampler.sample(bucket.alpha(), bucket.beta());
-
-    let mut buckets = buckets.iter_mut();
-
-    let first_bucket = buckets.next().ok_or(Error::NoBucketsToPull)?;
-    let first_sample = sample_from_bucket(first_bucket)?;
-
-    let bucket = buckets
-        .try_fold(
-            (first_sample, first_bucket),
-            |max, bucket| -> Result<_, Error> {
-                let sample = sample_from_bucket(bucket)?;
-                if let Ordering::Greater = nan_safe_f32_cmp(&sample, &max.0) {
-                    Ok((sample, bucket))
-                } else {
-                    Ok(max)
-                }
-            },
-        )?
-        .1;
-
-    bucket.pop().ok_or(Error::EmptyBucket)
+/// Samples the next bucket.
+fn pull_arms<T>(
+    beta_sampler: &impl BetaSample,
+    buckets: Vec<impl Deref<Target = impl Bucket<T>>>,
+) -> Option<Result<usize, Error>> {
+    match buckets
+        .into_iter()
+        .enumerate()
+        .filter(|(_, bucket)| !bucket.is_empty())
+        .try_fold(None, |max, (index, bucket)| {
+            beta_sampler
+                .sample(bucket.alpha(), bucket.beta())
+                .map(|sample| {
+                    if let Some((max_sample, _)) = max {
+                        if let Ordering::Greater = nan_safe_f32_cmp(&sample, &max_sample) {
+                            Some((sample, index))
+                        } else {
+                            max
+                        }
+                    } else {
+                        Some((sample, index))
+                    }
+                })
+        }) {
+        Ok(Some((_, index))) => Some(Ok(index)),
+        Ok(None) => None,
+        Err(error) => Some(Err(error)),
+    }
 }
 
-struct SelectionIter<'bs, 'b, BS, B, T>
-where
-    B: Bucket<T>,
-{
-    beta_sampler: &'bs BS,
-    buckets: Vec<&'b mut B>,
-    bucket_type: PhantomData<T>,
+/// A stream to select elements from buckets.
+pub(crate) struct Selection<'b, BS, B, T> {
+    beta_sampler: BS,
+    buckets: Vec<&'b RwLock<B>>,
+    bucket_type: PhantomData<&'b T>,
 }
 
-impl<'bs, 'b, BS, B, T> SelectionIter<'bs, 'b, BS, B, T>
-where
-    B: Bucket<T>,
-{
-    fn new(beta_sampler: &'bs BS, buckets: Vec<&'b mut B>) -> Self {
+impl<'b, BS, B, T> Selection<'b, BS, B, T> {
+    /// Creates a selective steam.
+    pub(crate) fn new(beta_sampler: BS, buckets: impl IntoIterator<Item = &'b RwLock<B>>) -> Self {
```

Comment:
Can't we remove `RwLock` from this module completely and leave to the user to job to make the data in a format that the mab module can use?

Reply:
Having:
```rust
 pub async fn get_feed_documents(&self, max_documents: usize) -> Result<Vec<Document>, Error> {
        let mut stacks = Vec::with_capacity(self.stacks.len());
        for stack in self.stacks.values() {
            let stack = stack.write().await;
            stacks.push(stack);
        }

        Selection::new(BetaSampler)
            .select(stacks, max_documents)
            .map_err(|e| e.into())
}
```
and w.r.t the mab.rs that we have in main, changing:
```rust
fn pull_arms<B, T>(
    beta_sampler: &impl BetaSample,
    buckets: &mut [impl DerefMut<Target = B>],
) -> Result<T, Error>
```

```rust
struct SelectionIter<'bs, BS, D, B, T>
where
    D: DerefMut<Target = B>,
    B: Bucket<T>,
{
    beta_sampler: &'bs BS,
    buckets: Vec<D>,
    bucket_type: PhantomData<T>,
}
```

```rust
struct SelectionIter<'bs, BS, D, B, T>
where
    D: DerefMut<Target = B>,
    B: Bucket<T>,
{
    beta_sampler: &'bs BS,
    buckets: Vec<D>,
    bucket_type: PhantomData<T>,
}
```
and some other adjustment that derives from this we can have the mab work without async code and minimal changes.

Reply:
my original idea was to have `stacks: Hashmap<StackId, RwLock<Stack>>` to allow concurrent operations on each stack, for example having a selection stream instead of a selection iter. i guess it makes sense to change it to `stacks: RwLock<HashMap<StackId, Stack>>` now, since we need to acquire all locks in many cases anyways. this will also make the code simpler/changes less, if we need more fine grained locks, we can introduce them again later on.

Reply:
What we would at the end is to be able to download the data from the api and run (part of?) the filters without locking the selection on the stack. For the mab we want to get all the `n` elements that are requested with the state we have when we start, so for this we need to lock all stacks first and get extract elements from that. I'm also fine with having an `RwLock` around the map at the moment and we can optimize it later. Also when we have the database this can also be removed and run the mab inside a transaction or something like that.